safe-global/web-core
# fix: Reset Core SDK when switching safes

## What it solves

Resolves #1253 

## How this PR fixes it

There are 2 ways of switching to a safe on a different network:

1. User is on Safe A
2. They switch to Safe B and change their wallet network from A to B
---
1. User is on Safe A
2. They change their wallet network to B from within their wallet
3. They switch to Safe B

In the second case we try to initialize the Core SDK twice, once with stale data from Safe A and the second time with the correct data from Safe B. The first call throws a `SafeProxy contract is not deployed on the current network` error. Depending on which promise resolves first it then resets the Core SDK instance which breaks the UI.

### Other fixes

- Hides the Send button on the assets page and address book if the wallet is on a wrong network. This is the same behaviour as for all other buttons that can result in a transaction

## How to test it

Go the the assets page and address book and switch network in your wallet. The Send button should disappear.

1. Open a Safe on Görli
2. Navigate to a Safe app and create a transaction
4. Observe the correct nonce being shown in the modal
5. Switch to a different network from your wallet
6. Observe the modal showing a `Your wallet is connected to the wrong chain.` error
7. Switch to a Safe on the new network
8. Repeat steps 2-7

---

1. Open a safe app on mainnet while the wallet is set to a different network
2. Reload the page
3. Try to create a transaction so that the modal opens
4. Switch the wallet networ to mainnet
5. Observe that no error is being displayed and the correct nonce is there

---

Additionally, we should do a full regression on all types of transactions (Batch execute, create/remove/update spending limit, add/remove/change owner, sign/sign+execute/execute tx, ...)
